python/cudf/cudf/tests/reshape/test_join.py

```diff
@@ -0,0 +1,670 @@
+    assert_exceptions_equal,
+)
+
+
+@pytest.fixture(
+    params=(
+        "left",
+        "inner",
+        "outer",
+        "right",
+        "leftanti",
+        "leftsemi",
+        "cross",
+    )
+)
+def how(request):
+    return request.param
+
+
+def assert_join_results_equal(expect, got, how, **kwargs):
+    if how == "right":
+        got = got[expect.columns]
+
+    if isinstance(expect, (pd.Series, cudf.Series)):
+        return assert_eq(
+            expect.sort_values().reset_index(drop=True),
+            got.sort_values().reset_index(drop=True),
+            **kwargs,
+        )
+    elif isinstance(expect, (pd.DataFrame, cudf.DataFrame)):
+        if not len(
+            expect.columns
+        ):  # can't sort_values() on a df without columns
+            return assert_eq(expect, got, **kwargs)
+
+        assert_eq(
+            expect.sort_values(expect.columns.to_list()).reset_index(
+                drop=True
+            ),
+            got.sort_values(got.columns.to_list()).reset_index(drop=True),
+            **kwargs,
+        )
+    elif isinstance(expect, (pd.Index, cudf.Index)):
+        return assert_eq(expect.sort_values(), got.sort_values(), **kwargs)
+    else:
+        raise ValueError(f"Not a join result: {type(expect).__name__}")
+
+
+@pytest.mark.parametrize(
+    "aa, bb",
+    [
+        [[0, 0, 4, 5, 5], [0, 0, 2, 3, 5]],
+        [[0, 0, 1, 2, 3], [0, 1, 2, 2, 3]],
+        [range(5), range(5, 10)],
+        [[0.1, 0.2, 0.3, 0.4, 0.5], [0.6, 0.7, 0.8, 0.9, 1.0]],
+    ],
+)
+def test_dataframe_join_how(aa, bb, how):
+    df = cudf.DataFrame(
+        {
+            "a": aa,
+            "b": bb,
+        }
+    )
+
+    def work_pandas(df, how):
+        df1 = df.set_index("a")
+        df2 = df.set_index("b")
+        if how == "leftanti":
+            joined = df1[~df1.index.isin(df2.index)][df1.columns]
+        elif how == "leftsemi":
+            joined = df1[df1.index.isin(df2.index)][df1.columns]
+        else:
+            joined = df1.join(df2, how=how, sort=True)
+        return joined
+
+    def work_gdf(df):
+        df1 = df.set_index("a")
+        df2 = df.set_index("b")
+        joined = df1.join(df2, how=how, sort=True)
+        return joined
+
+    expect = work_pandas(df.to_pandas(), how)
+    got = work_gdf(df)
+    expecto = expect.copy()
+    goto = got.copy()
+
+    expect = expect.astype(np.float64).fillna(np.nan)[expect.columns]
+    got = got.astype(np.float64).fillna(np.nan)[expect.columns]
+
+    assert got.index.name is None
+
+    assert list(expect.columns) == list(got.columns)
+    if how in {"left", "inner", "right", "leftanti", "leftsemi"}:
+        assert_eq(sorted(expect.index.values), sorted(got.index.values))
+        if how != "outer":
+            # Newly introduced ambiguous ValueError thrown when
+            # an index and column have the same name. Rename the
+            # index so sorts work.
+            # TODO: What is the less hacky way?
```

Comment:
I see some interesting comments in here that we should probably come back and investigate...